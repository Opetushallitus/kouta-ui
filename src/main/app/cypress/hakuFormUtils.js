import merge from 'lodash/merge';

import organisaatio from './data/organisaatio';
import {
  stubKoodistoRoute,
  stubHakemuspalveluLomakkeetRoute,
  stubOppijanumerorekisteriHenkiloRoute,
} from './utils';

export const stubHakuFormRoutes = ({ cy, organisaatioOid }) => {
  cy.server();

  cy.route({
    method: 'GET',
    url: `**/organisaatio-service/rest/organisaatio/v4/${organisaatioOid}**`,
    response: merge(organisaatio(), {
      oid: organisaatioOid,
    }),
  });

  cy.route({
    method: 'POST',
    url: '**/organisaatio-service/rest/organisaatio/v4/findbyoids',
    response: [
      merge(organisaatio(), {
        oid: organisaatioOid,
      }),
    ],
  });

  stubKoodistoRoute({ koodisto: 'haunkohdejoukko', cy });
  stubKoodistoRoute({ koodisto: 'hakutapa', cy });
  stubKoodistoRoute({ koodisto: 'kausi', cy });
  stubKoodistoRoute({ koodisto: 'opetuspaikkakk', cy });
  stubKoodistoRoute({ koodisto: 'kausi', cy });

  cy.route({
    method: 'GET',
    url: '**/haku/list**',
    response: [],
  });

  stubHakemuspalveluLomakkeetRoute({ cy });
  stubOppijanumerorekisteriHenkiloRoute({ cy });
};
