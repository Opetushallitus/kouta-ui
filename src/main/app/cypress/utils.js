import koodisto from '#/cypress/data/koodisto';

export const getByTestId = (testId, cy) => {
  return cy.get(`[data-testid="${testId}"]`);
};

export const getRadio = (value, cy) => {
  return cy.get(`input[type="radio"][value="${value}"]`);
};

export const getSelectOption = (value, cy) => {
  return cy.get('[class*="option"]').contains(value);
};

export const getCheckbox = (value, cy) => {
  return cy.get(`input[type="checkbox"]${value ? `[name="${value}"]` : ''}`);
};

export const getSelect = cy => {
  return cy.get('.Select__');
};

export const selectOption = (value, cy) => {
  getSelect(cy).click();

  getSelect(cy).within(() => {
    getSelectOption(value, cy).click({ force: true });
  });
};

export const fillAsyncSelect = (input, match) => {
  getSelect(cy).find('input[type="text"]').paste(input);

  getSelect(cy).within(() => {
    cy.get(`div:contains(${match})`).first().click();
  });
};

export const stubKoodistoRoute = ({ koodisto: koodistonNimi, cy }) => {
  cy.route({
    method: 'GET',
    url: `**/koodisto-service/rest/json/${koodistonNimi}/koodi**`,
    response: koodisto({ koodisto: koodistonNimi }),
  });
};

export const stubLokalisaatioRoute = ({ cy }) => {
  cy.route({
    method: 'GET',
    url: '**/lokalisointi/cxf/rest/v1/localisation**',
    response: [],
  });
};

export const typeToEditor = (value, cy) => {
  cy.get('.Editor__').within(() => {
    cy.get('[contenteditable="true"]').paste(value);
  });
};

export const getTableInput = cy => {
  return cy.get('.TableInput__');
};

export const fillDateTimeInput = ({ date, time, cy }) => {
  getByTestId('DateTimeInput', cy).within(() => {
    getByTestId('DateTimeInput__Date', cy).find('input').paste(date);
    getByTestId('DateTimeInput__Time', cy).find('input').clear().paste(time);
  });
};

export const chooseKieliversiotLanguages = (selectedLanguages, cy) => {
  const languages = ['en', 'fi', 'sv'];
  languages.forEach(lang => {
    cy.get(`input[name=${lang}]`).as('checkbox');

    if (selectedLanguages.includes(lang)) {
      cy.get('@checkbox').check({ force: true });
    } else {
      cy.get('@checkbox').uncheck({ force: true });
    }
  });
};

export const fillYhteyshenkilotFields = ({ cy }) => {
  cy.getByTestId('lisaaYhteyshenkiloButton').click({ force: true });

  cy.getByTestId('nimi').find('input').paste('nimi');
  cy.getByTestId('titteli').find('input').paste('titteli');
  cy.getByTestId('sahkoposti').find('input').paste('sÃ¤hkoposti');
  cy.getByTestId('puhelinnumero').find('input').paste('puhelin');
  cy.getByTestId('verkkosivu').find('input').paste('verkkosivu');
};

export const fillValintakoeFields = () => {
  selectOption('valintakokeentyyppi_1', cy);

  cy.getByTestId('lisaaTilaisuusButton').click({ force: true });

  cy.getByTestId('osoite').find('input').paste('osoite');

  cy.getByTestId('postinumero').within(() => {
    fillAsyncSelect('0', '0 Posti_0');
  });

  cy.getByTestId('alkaa').within(() => {
    fillDateTimeInput({
      date: '02.04.2019',
      time: '10:45',
      cy,
    });
  });

  cy.getByTestId('paattyy').within(() => {
    fillDateTimeInput({
      date: '02.04.2019',
      time: '19:00',
      cy,
    });
  });

  cy.getByTestId('lisatietoja').find('textarea').paste('lisatietoja');
};

export const stubKayttoOikeusMeRoute = ({ user = {}, cy }) => {
  cy.route({
    method: 'GET',
    url: '**/kayttooikeus-service/cas/me',
    response: {
      uid: 'johndoe',
      oid: '1.2.246.562.24.62301161440',
      firstName: 'John',
      lastName: 'Doe',
      lang: 'fi',
      roles: JSON.stringify([
        'APP_KOUTA_OPHPAAKAYTTAJA_1.2.246.562.10.00000000001',
      ]),
      ...user,
    },
  });
};

export const stubKoutaBackendLoginRoute = ({ cy }) => {
  cy.route({
    method: 'GET',
    url: '**/kouta-backend/auth/login',
    response: {},
  });
};

export const stubKoutaBackendSessionRoute = ({ cy }) => {
  cy.route({
    method: 'GET',
    url: '**/kouta-backend/auth/session',
    response: {},
  });
};

export const stubHakemuspalveluLomakkeetRoute = ({
  cy,
  lomakkeet = [{ name: { fi: 'Lomake 1' }, key: 'lomake_1' }],
}) => {
  cy.route({
    method: 'GET',
    url: '**/lomake-editori/api/forms',
    response: {
      forms: lomakkeet,
    },
  });
};

export const stubOppijanumerorekisteriHenkiloRoute = ({
  cy,
  henkilo = { etunimet: 'John', sukunimi: 'Doe' },
}) => {
  cy.route({
    method: 'GET',
    url: '**/oppijanumerorekisteri-service/henkilo/**',
    response: henkilo,
  });
};

export const fillTreeSelect = (value, cy) => {
  value.forEach(val => {
    cy.get(`input[type="checkbox"][name="${val}"]`).check({ force: true });
  });
};

export const fillKoulutustyyppiSelect = (path, cy) => {
  path.forEach(option => {
    getRadio(option, cy).check({ force: true });
  });
};

export const fillDatePickerInput = value => {
  cy.get('.DatePickerInput__').find('input').paste(value);
};

export const stubKoodiRoute = koodi => {
  const { koodiUri, versio } = koodi;

  return cy.route({
    method: 'GET',
    url: `/koodisto-service/rest/codeelement/${koodiUri}/${versio}`,
    response: koodi,
  });
};

export const stubEPerusteetByKoulutuskoodiRoute = () => {
  cy.route({
    method: 'GET',
    url:
      '**/eperusteet-service/api/perusteet?tuleva=true&siirtyma=false&voimassaolo=true&poistunut=false&kieli=fi&koulutuskoodi=koulutus_0',
    response: {
      data: [
        {
          nimi: { fi: 'koulutus_0' },
          kuvaus: { fi: 'koulutus_0 kuvaus' },
          osaamisalat: [
            { uri: 'osaamisala_0', nimi: { fi: 'osaamisala_0 nimi' } },
          ],
          tutkintonimikeKoodiUri: 'nimike_1#1',
          id: 1,
        },
      ],
    },
  });
};

export const stubCommonRoutes = () => {
  stubLokalisaatioRoute({ cy });
  stubKayttoOikeusMeRoute({ cy });
  stubKoutaBackendLoginRoute({ cy });
  stubKoutaBackendSessionRoute({ cy });
};

export const jatka = () => cy.getByTestId('jatkaButton').click({ force: true });

export const OPH_TEST_ORGANISAATIO_OID = '1.2.246.562.10.48587687889';

export const isStubbed =
  !process.env.CYPRESS_BACKEND || process.env.CYPRESS_BACKEND === 'stubs';
