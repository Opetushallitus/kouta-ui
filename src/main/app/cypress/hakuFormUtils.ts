import { playMocks } from 'kto-ui-common/cypress/mockUtils';
import { merge } from 'lodash/fp';

import organisaatio from '#/cypress/data/organisaatio';
import organisaatioHierarkia from '#/cypress/data/organisaatioHierarkia';
import hakuMocks from '#/cypress/mocks/haku.mocks.json';

import {
  stubHakemuspalveluLomakkeetRoute,
  stubOppijanumerorekisteriHenkiloRoute,
  stubCommonRoutes,
} from './utils';

export const stubHakuFormRoutes = ({ organisaatioOid }) => {
  stubCommonRoutes();
  playMocks(hakuMocks);

  cy.intercept(
    {
      method: 'GET',
      url: `**/organisaatio-service/rest/organisaatio/v4/hierarkia/hae**oid=${organisaatioOid}**`,
    },
    { body: organisaatioHierarkia({ rootOid: organisaatioOid }) }
  );

  cy.intercept(
    {
      method: 'GET',
      url: `**/organisaatio-service/rest/organisaatio/v4/hierarkia/hae**`,
    },
    { body: organisaatioHierarkia({ rootOid: organisaatioOid }) }
  );

  cy.intercept(
    {
      method: 'GET',
      url: `**/organisaatio-service/rest/organisaatio/v4/${organisaatioOid}**`,
    },
    {
      body: merge(organisaatio(), {
        oid: organisaatioOid,
      }),
    }
  );

  cy.intercept(
    {
      method: 'POST',
      url: '**/organisaatio-service/rest/organisaatio/v4/findbyoids',
    },
    {
      body: [
        merge(organisaatio(), {
          oid: organisaatioOid,
        }),
      ],
    }
  );

  cy.intercept({ method: 'GET', url: '**/haku/list**' }, { body: [] });

  cy.intercept({ method: 'GET', url: `**/toteutus/list**` }, { body: [] });

  cy.intercept({ method: 'GET', url: `**/koulutus/list**` }, { body: [] });

  cy.intercept({ method: 'GET', url: `**/search/haku/**` }, { body: [] });

  stubHakemuspalveluLomakkeetRoute();
  stubOppijanumerorekisteriHenkiloRoute();
};
