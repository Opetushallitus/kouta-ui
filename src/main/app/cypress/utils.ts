import { fireEvent } from '@testing-library/react';
import { loggable } from 'cypress-pipe';
import { playMocks } from 'kto-ui-common/cypress/mockUtils';
import { isEmpty, includes, last, merge, toLower, uniqueId } from 'lodash/fp';

import commonMocks from '#/cypress/mocks/common.mocks.json';
import { Alkamiskausityyppi } from '#/src/constants';

export const paste = loggable('paste', value => $element => {
  $element.focus();
  try {
    fireEvent.change($element[0], { target: { value } });
  } catch (e) {
    cy.wrap($element).then($destination => {
      const pasteEvent = Object.assign(
        new Event('paste', { bubbles: true, cancelable: true }),
        {
          clipboardData: {
            getData: (type = 'text') => value,
          },
        }
      );
      $destination[0].dispatchEvent(pasteEvent);
    });
  }
  return cy.wrap($element);
});

export const getSelectOption = value =>
  cy.findAllByRole('option').contains(value);

export const pFillSelect = loggable(
  'fillSelect',
  value => $element =>
    cy
      .wrap($element)
      .click()
      .within(() => {
        getSelectOption(value).click({ force: true });
      })
);

export const pFillAsyncSelect = loggable(
  'fillAsyncSelect',
  (input, match = null) =>
    $element =>
      cy.wrap($element).within(() => {
        cy.get('input[type="text"]')
          .should('not.be.disabled')
          .pipe(paste(input));
        cy.findAllByRole('option', { name: includes(match || input) })
          .first()
          .click({ force: true });
      })
);

export const pFillEditor = loggable(
  'fillEditor',
  value => $element =>
    cy.wrap($element).within(() => {
      // NOTE: .clear wont work here -> use type instead
      cy.get('[contenteditable="true"]').type('{selectall}').pipe(paste(value));
    })
);

export const getInputByLabel = text =>
  cy.findByLabelText(text, { exact: false });

export const getSelectByLabel = text =>
  getInputByLabel(text).closest('.Select__');

export const getEditorByLabel = text =>
  getInputByLabel(text).closest('.Editor__');

export const getByTestId = cy.findByTestId;

export const getFieldWrapperByName = name =>
  cy.findByTestId(`form-control_${name}`);

export const getRadio = value =>
  cy.get(`input[type="radio"][value="${value}"]`);

export const getCheckbox = value =>
  cy.get(`input[type="checkbox"]${value ? `[name="${value}"]` : ''}`);

export const selectCheckbox = name =>
  cy.findByRole('checkbox', { name }).check({ force: true });

export const getSelect = () => cy.get('.Select__');

export const selectOption = value => getSelect().pipe(pFillSelect(value));

export const fillAsyncSelect = (input, match = null) => {
  getSelect().pipe(pFillAsyncSelect(input, match));
};

export const withinSection = (name, fn) =>
  cy.findByTestId(`${name}Section`).within($sectionEl => {
    const sectionIsOpen = !isEmpty($sectionEl.children('[open]'));

    if (!sectionIsOpen) {
      cy.wrap($sectionEl).click();
    }

    cy.wrap($sectionEl).within(() => {
      fn();
    });
  });

export const stubLokalisaatioRoute = () => {
  cy.intercept(
    { method: 'GET', url: '/lokalisointi/cxf/rest/v1/localisation' },
    { body: [] }
  );
};

export const typeToEditor = value =>
  cy.get('.Editor__').pipe(pFillEditor(value));

export const getTableInput = () => {
  return cy.get('.TableInput__');
};

export const fillDateTimeInput = ({ date, time }) => {
  getByTestId('DateTimeInput').within(() => {
    getByTestId('DateTimeInput__Date').find('input').clear().pipe(paste(date));
    getByTestId('DateTimeInput__Time').find('input').clear().pipe(paste(time));
  });
};

export const chooseKieliversiotLanguages = (selectedLanguages = []) => {
  const languages = ['en', 'fi', 'sv'];
  languages.forEach(lang => {
    cy.get(`input[name=${lang}]`)
      .should('not.be.disabled')
      .then($checkbox => {
        return selectedLanguages.includes(lang)
          ? cy.wrap($checkbox).check({ force: true })
          : cy.wrap($checkbox).uncheck({ force: true });
      });
  });
};

export const fillYhteyshenkilotFields = () => {
  cy.findByRole('button', { name: 'yleiset.lisaaYhteyshenkilo' }).click({
    force: true,
  });
  cy.findByRole('textbox', { name: 'yleiset.nimi' }).pipe(paste('nimi'));
  cy.findByRole('textbox', { name: 'yleiset.titteli' }).pipe(paste('titteli'));
  cy.findByRole('textbox', { name: 'yleiset.sahkoposti' }).pipe(
    paste('sähköposti')
  );

  cy.findByRole('textbox', { name: 'yleiset.puhelinnumero' }).pipe(
    paste('puhelin')
  );
  cy.findByRole('textbox', { name: 'yleiset.verkkosivu' }).pipe(
    paste('verkkosivu')
  );
};

export const stubKayttoOikeusMeRoute = ({ user = {} } = {}) => {
  cy.intercept(
    { method: 'GET', url: '/kayttooikeus-service/cas/me' },
    {
      body: {
        uid: 'johndoe',
        oid: '1.2.246.562.24.62301161440',
        firstName: 'John',
        lastName: 'Doe',
        lang: 'fi',
        roles: JSON.stringify([
          'APP_KOUTA',
          'APP_KOUTA_OPHPAAKAYTTAJA',
          'APP_KOUTA_OPHPAAKAYTTAJA_1.2.246.562.10.00000000001',
        ]),
        ...user,
      },
    }
  );
};

export const stubAsiointikieliRoute = () => {
  cy.intercept(
    {
      method: 'GET',
      url: '/oppijanumerorekisteri-service/henkilo/current/asiointiKieli',
    },
    {
      body: 'fi',
    }
  );
};

export const stubKoutaBackendLoginRoute = () => {
  cy.intercept(
    { method: 'GET', url: '/kouta-backend/auth/login' },
    { statusCode: 200, body: {} }
  );
};

export const stubKoutaBackendSessionRoute = () => {
  cy.intercept(
    { method: 'GET', url: '/kouta-backend/auth/session' },
    { statusCode: 200, body: {} }
  );
};

export const stubHakemuspalveluLomakkeetRoute = ({
  lomakkeet = [{ name: { fi: 'Lomake 1' }, key: 'lomake_1' }],
} = {}) => {
  cy.intercept(
    { method: 'GET', url: '/lomake-editori/api/forms' },
    {
      body: {
        forms: lomakkeet,
      },
    }
  );
};

export const stubOppijanumerorekisteriHenkiloRoute = ({
  henkilo = { etunimet: 'John', sukunimi: 'Doe' },
} = {}) => {
  cy.intercept(
    { method: 'GET', url: '/oppijanumerorekisteri-service/henkilo/' },
    { body: henkilo }
  );
};

export const fillTreeSelect = value => {
  value.forEach(val => {
    cy.get(`input[type="checkbox"][name="${val}"]`).check({ force: true });
  });
};

export const fillDatePickerInput = value => {
  cy.get('.DatePickerInput__').find('input').pipe(paste(value));
};

const koutaSearchItem = ({ idProp = 'oid' } = { idProp: 'oid' }) => ({
  modified: '2019-02-20T07:55',
  muokkaaja: { nimi: 'John Doe', oid: '1.2.246.562.24.62301161440' },
  nimi: { fi: 'Nimi' },
  [idProp]: '1.1.1.1.1.1',
  organisaatio: {},
  tila: 'tallennettu',
});

const stubEntityLists = () => {
  cy.intercept(
    { method: 'GET', url: '/kouta-backend/search/koulutukset?*' },
    {
      body: {
        result: [
          merge(koutaSearchItem(), { nimi: { fi: 'Koulutuksen nimi' } }),
        ],
        totalCount: 1,
      },
    }
  );

  cy.intercept(
    { method: 'GET', url: '/kouta-backend/search/toteutukset?*' },
    {
      body: {
        result: [
          merge(koutaSearchItem(), { nimi: { fi: 'Toteutuksen nimi' } }),
        ],
        totalCount: 1,
      },
    }
  );

  cy.intercept(
    { method: 'GET', url: '/kouta-backend/search/haut?*' },
    {
      body: {
        result: [merge(koutaSearchItem(), { nimi: { fi: 'Haun nimi' } })],
        totalCount: 1,
      },
    }
  );

  cy.intercept(
    { method: 'GET', url: '/kouta-backend/search/valintaperusteet?*' },
    {
      body: {
        result: [
          merge(koutaSearchItem({ idProp: 'id' }), {
            nimi: { fi: 'Valintaperusteen nimi' },
          }),
        ],
        totalCount: 1,
      },
    }
  );

  cy.intercept(
    { method: 'GET', url: '/kouta-backend/search/hakukohteet?*' },
    {
      body: {
        result: [
          merge(koutaSearchItem(), { nimi: { fi: 'Hakukohteen nimi' } }),
        ],
        totalCount: 1,
      },
    }
  );
};

export const stubCommonRoutes = () => {
  stubEntityLists();
  stubLokalisaatioRoute();
  stubKayttoOikeusMeRoute();
  stubAsiointikieliRoute();
  stubKoutaBackendLoginRoute();
  stubKoutaBackendSessionRoute();
  playMocks(commonMocks);
};

export const jatka = () =>
  cy.findByRole('button', { name: 'yleiset.jatka' }).click();

export const OPH_TEST_ORGANISAATIO_OID = '1.2.246.562.10.48587687889';

export const isStubbed =
  !process.env.CYPRESS_BACKEND || process.env.CYPRESS_BACKEND === 'stubs';

const fillTilaisuus = () => {
  getByTestId('lisaaTilaisuusButton').click({ force: true });
  getByTestId('osoite').find('input').pipe(paste('osoite'));
  getByTestId('postinumero').within(() => {
    fillAsyncSelect('00350');
  });
  getByTestId('alkaa').within(() => {
    fillDateTimeInput({
      date: '02.04.2019',
      time: '10:45',
    });
  });

  getByTestId('paattyy').within(() => {
    fillDateTimeInput({
      date: '02.04.2019',
      time: '19:00',
    });
  });

  getByTestId('jarjestamispaikka').find('input').pipe(paste('paikka'));
  getByTestId('lisatietoja').within(() => {
    typeToEditor('lisatietoja');
  });
};

export const fillValintakokeetSection = ({
  withValintaperusteenKokeet = false,
} = {}) => {
  withinSection('valintakokeet', () => {
    getByTestId('yleisKuvaus').within(() => {
      typeToEditor('Valintakokeiden kuvaus');
    });

    if (withValintaperusteenKokeet) {
      getByTestId('valintaperusteenValintakokeet').within(() => {
        fillTilaisuus();
      });
    }

    getByTestId('kokeetTaiLisanaytot').within(() => {
      getByTestId('lisaaKoeTaiLisanayttoButton').click({ force: true });
      getByTestId('kokeenTaiLisanaytonTyyppi').within(() => {
        selectOption('Valintakoe');
      });
      getByTestId('hakijalleNakyvaNimi').find('input').pipe(paste('nimi'));

      getByTestId('tietoaHakijalle').within(() => {
        typeToEditor('Tietoa hakijalle');
      });

      getByTestId('vahimmaispistemaara').find('input').pipe(paste('10,03'));

      getByTestId('liittyyEnnakkovalmistautumista').within(() => {
        getCheckbox(null).check({ force: true });
      });

      getByTestId('ohjeetEnnakkovalmistautumiseen').within(() => {
        typeToEditor('ohjeet ennakkovalmistautumiseen');
      });

      getByTestId('erityisjarjestelytMahdollisia').within(() => {
        getCheckbox(null).check({ force: true });
      });

      getByTestId('ohjeetErityisjarjestelyihin').within(() => {
        typeToEditor('ohjeet erityisjärjestelyihin');
      });

      fillTilaisuus();
    });
    jatka();
  });
};

export const fillKieliversiotSection = (
  { jatka: pressJatka } = {
    jatka: false,
  }
) => {
  withinSection('kieliversiot', () => {
    chooseKieliversiotLanguages(['fi']);
    if (pressJatka) {
      jatka();
    }
  });
};

export const fillPohjaSection = () => {
  withinSection('pohja', () => {
    jatka();
  });
};

export const fillTilaSection = (tila = 'julkaistu') => {
  withinSection('tila', () => {
    getRadio(tila).check({ force: true });
  });
};

export const tallenna = () => {
  cy.findByRole('button', { name: 'yleiset.tallenna' }).click();
};

const isTutkintoonJohtava = koulutustyyppi =>
  ['amk', 'yo', 'amm', 'lk'].includes(koulutustyyppi);

export const fillKoulutustyyppiSelect = koulutustyyppiPath => {
  const johtaaTutkintoon = isTutkintoonJohtava(last(koulutustyyppiPath));

  if (johtaaTutkintoon) {
    cy.findByRole('button', {
      name: 'koulutustyyppivalikko.tutkintoonJohtavatKoulutustyypit',
    }).click();
  } else {
    cy.findByRole('button', {
      name: 'koulutustyyppivalikko.muutKoulutustyypit',
    }).click();
  }

  koulutustyyppiPath.forEach(option => {
    getRadio(option).check({ force: true });
  });
};

export const fillKoulutustyyppiSection = koulutustyyppiPath => {
  withinSection('koulutustyyppi', () => {
    fillKoulutustyyppiSelect(koulutustyyppiPath);
    jatka();
  });
};

export const assertNoUnsavedChangesDialog = () => {
  cy.findByRole('link', { name: /home/i }).click();
  cy.findByRole('heading', {
    name: 'ilmoitukset.tallentamattomiaMuutoksia.otsikko',
  }).should('not.exist');
};

export const fillAjankohtaFields = (
  alkamiskausityyppi = Alkamiskausityyppi.ALKAMISKAUSI_JA_VUOSI
) => {
  getByTestId('AloitusajankohtaFields').within(() => {
    switch (alkamiskausityyppi) {
      case Alkamiskausityyppi.ALKAMISKAUSI_JA_VUOSI:
        cy.findByText('yleiset.alkamiskausiJaVuosi').click();

        getRadio('kausi_k#1').click({ force: true });
        selectOption(2035);
        return;
      case Alkamiskausityyppi.TARKKA_ALKAMISAJANKOHTA:
        cy.findByText('yleiset.tarkkaAlkamisajankohta').click();

        getByTestId('alkaa').within(() => {
          fillDateTimeInput({
            date: '1.11.2030',
            time: '00:00',
          });
        });

        getByTestId('paattyy').within(() => {
          fillDateTimeInput({
            date: '30.11.2030',
            time: '00:00',
          });
        });
        return;
      case Alkamiskausityyppi.HENKILOKOHTAINEN_SUUNNITELMA:
        cy.findByText(
          'yleiset.aloitusHenkilokohtaisenSuunnitelmanMukaisesti'
        ).click();
        typeToEditor('Henkilökohtaisen suunnitelman lisätiedot');
    }
  });
};

/* Curried funktio, joka toteuttaa create ja edit testien yhteiset osat (tallennus-pyynnön vertaaminen snapshottiin).
 * "run" on funktio, joka ajaa muun testikoodin
 * options, objektissa entity on ENTITY-enum, jolle testi luodaan. oid/id on entiteetin tunniste
 * ("id" SORA-kuvauksella ja valintaperusteella)
 */
export const wrapMutationTest = options => run => () => {
  const { entity, oid, id } = options;

  const entityLower = toLower(entity);
  const requestAlias = `${entityLower}${uniqueId('Request')}`;

  cy.intercept(
    {
      method: '{PUT,POST}',
      url: `**/kouta-backend/${entityLower}`,
    },
    { body: oid ? { oid } : { id } }
  ).as(requestAlias);

  run();

  cy.wait(`@${requestAlias}`).then(({ request }) => {
    cy.wrap(request.body).toMatchSnapshot();
  });
};
