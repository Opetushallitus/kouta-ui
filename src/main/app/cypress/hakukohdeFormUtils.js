import merge from 'lodash/merge';

import { stubKoodistoRoute } from './utils';
import organisaatio from './data/organisaatio';
import haku from './data/haku';

export const stubHakukohdeFormRoutes = ({ cy, organisaatioOid, hakuOid }) => {
  cy.server();

  cy.route({
    method: 'GET',
    url: `**/organisaatio-service/rest/organisaatio/v4/${organisaatioOid}**`,
    response: merge(organisaatio(), {
      oid: organisaatioOid,
    }),
  });

  stubKoodistoRoute({ koodisto: 'pohjakoulutusvaatimustoinenaste', cy });
  stubKoodistoRoute({ koodisto: 'kausi', cy });
  stubKoodistoRoute({ koodisto: 'valintakokeentyyppi', cy });
  stubKoodistoRoute({ koodisto: 'liitetyypitamm', cy });

  cy.route({
    method: 'GET',
    url: `**/haku/${hakuOid}`,
    response: merge(haku(), {
      oid: hakuOid,
      organisaatioOid: organisaatioOid,
    }),
  });

  cy.route({
    method: 'GET',
    url: `**/hakukohde/list**`,
    response: [],
  });
};
