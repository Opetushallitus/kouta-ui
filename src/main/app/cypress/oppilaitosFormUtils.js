import merge from 'lodash/merge';

import createKoodi from './data/koodi';
import createOrganisaatio from './data/organisaatio';
import createOrganisaatioHierarkia from './data/organisaatioHierarkia';
import { stubKoodiRoute, stubKoodistoRoute } from './utils';

export const stubOppilaitosFormRoutes = ({ organisaatioOid }) => {
  stubKoodiRoute(createKoodi({ koodisto: 'posti', versio: 2 }));

  stubKoodistoRoute({ koodisto: 'koulutuksenlisatiedot', cy });

  cy.route({
    method: 'POST',
    url: '**/organisaatio-service/rest/organisaatio/v4/findbyoids',
    response: [
      merge(createOrganisaatio(), {
        oid: organisaatioOid,
      }),
    ],
  });

  cy.route({
    method: 'GET',
    url: `**/organisaatio-service/rest/organisaatio/v4/hierarkia/hae?oid=${organisaatioOid}**`,
    response: createOrganisaatioHierarkia({ rootOid: organisaatioOid }),
  });
};
