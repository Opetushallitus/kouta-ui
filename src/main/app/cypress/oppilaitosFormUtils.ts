import { playMocks } from 'kto-ui-common/cypress/mockUtils';
import { merge } from 'lodash/fp';

import createOrganisaatio from '#/cypress/data/organisaatio';
import createOrganisaatioHierarkia from '#/cypress/data/organisaatioHierarkia';
import oppilaitosMocks from '#/cypress/mocks/oppilaitos.mocks.json';
import { stubCommonRoutes } from '#/cypress/utils';

export const stubOppilaitosFormRoutes = ({ organisaatioOid }) => {
  stubCommonRoutes();
  playMocks(oppilaitosMocks);

  cy.intercept(
    {
      method: 'POST',
      url: '**/organisaatio-service/rest/organisaatio/v4/findbyoids',
    },
    {
      body: [
        merge(createOrganisaatio(), {
          oid: organisaatioOid,
        }),
      ],
    }
  );

  cy.intercept(
    {
      method: 'GET',
      url: `**/organisaatio-service/rest/organisaatio/v4/hierarkia/hae**oid=${organisaatioOid}**`,
    },
    { body: createOrganisaatioHierarkia({ rootOid: organisaatioOid }) }
  );
};
