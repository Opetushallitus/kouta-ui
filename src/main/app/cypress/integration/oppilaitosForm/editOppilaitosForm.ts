import createOppilaitos from '#/cypress/data/oppilaitos';
import { getByTestId, fillKieliversiotSection } from '#/cypress/utils';
import { stubOppilaitosFormRoutes } from '#/cypress/oppilaitosFormUtils';

const tallenna = () => {
  getByTestId('tallennaOppilaitosButton').click();
};

describe('editOppilaitosForm', () => {
  const organisaatioOid = '1.1.1.1.1.1';

  const oppilaitos = {
    ...createOppilaitos(),
    oid: organisaatioOid,
    organisaatioOid,
  };

  beforeEach(() => {
    cy.server();
    stubOppilaitosFormRoutes({ organisaatioOid });

    cy.route({
      method: 'GET',
      url: `**/oppilaitos/${oppilaitos.oid}`,
      response: oppilaitos,
    });

    cy.visit(`/organisaatio/${organisaatioOid}/oppilaitos`);
  });

  it('should be able to edit oppilaitos', () => {
    cy.route({
      method: 'POST',
      url: '**/oppilaitos',
      response: {},
    }).as('editOppilaitosResponse');

    fillKieliversiotSection();

    tallenna();

    cy.wait('@editOppilaitosResponse').then(({ request }) => {
      cy.wrap(request.body).toMatchSnapshot();
    });
  });
});
