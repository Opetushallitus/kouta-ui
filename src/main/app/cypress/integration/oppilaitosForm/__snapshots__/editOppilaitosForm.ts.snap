exports[`editOppilaitosForm > should be able to edit oppilaitos #0`] =
{
  "kielivalinta": [
    "fi"
  ],
  "logo": null,
  "metadata": {
    "akatemioita": 1,
    "esittely": {
      "fi": "<p><em>Fi esittely</em></p>"
    },
    "kampuksia": 3,
    "korkeakouluja": 5,
    "opiskelijoita": 100,
    "tiedekuntia": 4,
    "tietoaOpiskelusta": [
      {
        "otsikkoKoodiUri": "organisaationkuvaustiedot_0#1",
        "teksti": {
          "fi": "Fi tiedot"
        }
      }
    ],
    "toimipisteita": 9,
    "yhteystiedot": {
      "osoite": {
        "osoite": {
          "fi": "Fi osoite"
        },
        "postinumeroKoodiUri": "posti_0#2"
      },
      "puhelinnumero": {
        "fi": "1234"
      },
      "sahkoposti": {
        "fi": "fi@sahkoposti.fi"
      },
      "wwwSivu": {
        "fi": "www.verkkosivu.fi"
      }
    },
    "yksikoita": 2
  },
  "muokkaaja": "1.2.246.562.24.62301161440",
  "oid": "1.1.1.1.1.1",
  "organisaatioOid": "1.1.1.1.1.1",
  "teemakuva": null,
  "tila": "tallennettu"
};
